pricefeed/pricefeed.go

```diff
@@ -0,0 +1,143 @@
+package pricefeed
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/cockroachdb/apd/v3"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	logging "github.com/ipfs/go-log"
+
+	"github.com/athanorlabs/atomic-swap/common"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+)
+
+const (
+	// mainnetEndpoint is a mainnet ethereum endpoint, from
+	// https://chainlist.org/chain/1, which stagenet users get pointed at for price
+	// feeds, as Goeri doesn't have an XMR feed. Mainnet users will use the same
+	// ethereum endpoint that they use for other swap transactions.
+	mainnetEndpoint = "https://eth-rpc.gateway.pokt.network"
+
+	// https://data.chain.link/ethereum/mainnet/crypto-usd/eth-usd
+	chainlinkETHToUSDProxy = "0x5f4ec3df9cbd43714fe2740f5e3616155c5b8419"
+
+	// https://data.chain.link/ethereum/mainnet/crypto-usd/xmr-usd
+	chainlinkXMRToUSDProxy = "0xfa66458cce7dd15d8650015c4fce4d278271618f"
+)
+
+var (
+	errUnsupportedNetwork = errors.New("unsupported network")
+	log                   = logging.Logger("pricefeed")
+)
+
+// PriceFeed contains the interesting data from a chainlink price feed query.
+type PriceFeed struct {
+	Description string // "COIN / USD"
+	Price       *apd.Decimal
+	UpdatedAt   time.Time
+}
+
+// GetETHUSDPrice returns the current ETH/USD price from the Chainlink oracle.
+// It errors if the chain ID is not the Ethereum mainnet.
+func GetETHUSDPrice(ctx context.Context, ec *ethclient.Client) (*PriceFeed, error) {
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch chainID.Uint64() {
+	case common.MainnetChainID:
+		// No extra work to do
+	case common.GoerliChainID:
+		// Push stagenet/goerli users to a mainnet endpoint
+		ec, err = ethclient.Dial(mainnetEndpoint)
```

Comment:
why? just to make it more like mainnet?

Reply:
> why?

Because we want people testing the full functionality, and stagenet is the only option at the current time. If you try to start `swapd` with a mainnet endpoint, it will fail, because we don't have a SwapFactory contract deployed on mainnet yet.

> just to make it more like mainnet?

I think there is value in making the stagenet experience more or less identical to the mainnet experience, but without real money at stake. Stagenet is where they practice before going and doing the real thing.